### ISSUE TITLE
redis version 3.3.11 is too old, 3.4 or later required - what should I do?

### ISSUE THREAD

**Issue by wds4n6** [log2timeline/plaso#3163]
Had the following problem with 20200630 and 20200717.

Running SIFT with Ubuntu 18.04.4 LTS
Command:  log2timeline.py plaso.dump /home/sansforensics/Desktop/cases/case#/evidence#/evidence imageset.E01
Output was: 
2020-07-20 05:28:48.925 [INFO] (MainProcess) PID:53288 <data_location> Determined data location /usr/share/plaso
2020-07-20 05:28:48.925 [INFO] (MainProcess) PID:53288 <artifact _definitions> Determined artifact definitions path: /usr/share/artifacts
Checking availability and versions of dependencies.
[FAILURE]    redis version 3.3.11 is too old, 3.4 or later required.

Any assistance would be appreciated.
wds4n6@gmail.com

**Comment by joachimmetz** [comment]
You'll need to update your python3-redis version. There is a more recent version in the GIFT PPA. Unclear why SIFT does not provide a more recent version.

**Comment by joachimmetz** [comment]
Assuming question has been answered, closing issue.

**Comment by fergusond88** [comment]
running this worked for me:
sudo pip3 install redis

**Comment by joachimmetz** [comment]
@fergusond88 don't give bad advice 

https://plaso.readthedocs.io/en/latest/sources/developer/Developing-Virtualenv.html

**Comment by fergusond88** [comment]
@joachimmetz agreed. I should point out that I run L2T on a dedicated box that only does L2T and Timesketch. Error via omission :). Thanks for pointing that our

**Comment by honorthecode** [comment]
@joachimmetz , could you privide any tips on how to update the redis version?

I saw that you mentioned that the GIFT PPA has a more recent version of python3-redis so I ran the following:

1. sudo add-apt-repository ppa:gift/stable.

**Comment by joachimmetz** [comment]
@honorthecode can you explain your set up a bit, since I read "ELK container", assuming that you use Docker somewhere in your set up.

First we strongly recommend using the Docker container, especially if you're not familiar with/or have the time for troubleshooting installation issues.

1. Where are you running Plaso (in or outside the container)?
2. Where are you running the apt-get update?
3. Do you possibly have multiple versions of Pyhon and/or a second installation of Plaso on your system e.g. due to using pip previously?

We continuously test the installation from the PPA on Ubuntu 18.04 https://travis-ci.com/github/log2timeline/l2tbinaries/jobs/378728196

There are some troubleshooting tips mentioned here https://plaso.readthedocs.io/en/latest/sources/user/Troubleshooting-installation-issues.html. Sometimes your local configuration can prevent packages from being updated.

> I tried these steps both from a SIFT VM I download and also from a SIFT-CLI installation on a plain Ubuntu desktop install.

We don't support SIFT at the moment https://plaso.readthedocs.io/en/latest/sources/user/Ubuntu-Packaged-Release.html#sans-investigative-forensic-toolkit-sift-workstation. The SIFT maintainer recently indicated 18.04 is the default VM now in one of the issues but appears 16.04 is still mentioned in some of the documentation, so unclear to me what the status of SIFT is there.

**Comment by honorthecode** [comment]
Thanks for responding @joachimmetz . I have setups and encountered the redis problem in both.

Setup 1 is a Ubuntu plain 18.04 desktop where I installed the SIFT CLI and also the GIFT PPA in an effort to resolve the redis problem. On this system I am running ELK in a Docker container.

Setup 2 is a the most recent SIFT VM, which I can confirm is built on 18.04 (I agree that the references to 16.04 in some of the documentation is confusing). 
![image](https://user-images.githubusercontent.com/69635449/91628183-5d1b0800-e98b-11ea-89b9-438edc13155f.png)


1. Setup 1 and setup 2 were both setup this month for comparison and In both cases I've been running the Plaso installed directly on the systems rather than Plaso from a container. 
2. I'm not sure how to answer where I'm running apt-get update... I'm just running it from the Ubuntu OS terminal.
3. On both system I purposely avoided using pip because I have had problems with versions and dependencies in the past that I did not want to deal with this time. Setups 1 and 2 both seem to have Python 2 and 3 installed on it. For Setup 1, that may be the result of doing the SIFT-CLI installation.


I'm not opposed to using Plaso from a container, instead of the native installations. Do you know if that container has all the extras that allow L2T to parse out additional fields from artifacts such as security events. 
See below, the json on the left was created using Plaso a plain Ubuntu install with the GIFT PPA. The json on the right was using the SIFT VM using the same version of Plaso but more data was parsed out.

![image](https://user-images.githubusercontent.com/69635449/91628360-ea129100-e98c-11ea-9fe7-a3c3b685c73b.png)

Basically the whole goal I'm trying to achieve is to be able to ingest plaso data into ELK with as many fields parsed out as possible. If I can do that using containers for everything, I'm more than happy to do so.

**Comment by joachimmetz** [comment]
> I'm not sure how to answer where I'm running apt-get update... I'm just running it from the Ubuntu OS terminal.

So Plaso is installed on the same system (not in a container) where you are running apt-get

Maybe SIFT does something with your apt configuration. You can try using `aptitude versions python3-redis` do determine which versions of Python Redis are available to your configuration. 

> Do you know if that container has all the extras that allow L2T to parse out additional fields from artifacts such as security events.

I'm unfamiliar with these "extras". We don't include extras. If you want additional functionality in Plaso they need to be integrated following the code contribution process. 

> Basically the whole goal I'm trying to achieve is to be able to ingest plaso data into ELK with as many fields parsed out as possible.

The behavior you describe has been deprecated. The xml string contains the same information. This was never a use case mentioned before. We are considering to move this to the output instead. Also see: https://github.com/log2timeline/plaso/issues/2940

> Basically the whole goal I'm trying to achieve is to be able to ingest plaso data into ELK with as many fields parsed out as possible.

Be warned that the use of some of these "fields" is not necessarily "defined". Event log entries can contain data that is not used in the formatted message. It is unknown what the behavior of these fields are.

**Comment by honorthecode** [comment]
Very strange... I installed aptitude on the SIFT and ran `aptitude versions python3-redis`
It shows the following and still fails.
![image](https://user-images.githubusercontent.com/69635449/91640714-9be1aa00-e9ed-11ea-992c-6f74ce4bec38.png)

I have some time today so I'll try the Plaso container and see how that works.

**Comment by joachimmetz** [comment]
are you sure that your installation of Plaso uses your system installation of python3-redis and not that you have multiple installations in your PYTHONPATH, e.g. due to use of pip. (This is something we see very often)

See https://plaso.readthedocs.io/en/latest/sources/user/Troubleshooting-installation-issues.html for some troubleshooting tips.

**Comment by honorthecode** [comment]
Thanks @joachimmetz . I took a look at the troubleshooting page and ran ipython to learn a little more. 
![image](https://user-images.githubusercontent.com/69635449/91644963-11aa3d80-ea0f-11ea-9c63-8520ca9d4b5a.png)
 
The two things I notice here are that its identifying Python 2.7.17  and that after I imported redis the location was identified as being under /usr/local/lib/python2.7/...

To be certain I hadn't caused the problem somewhere along the way I reverted my Ubuntu Desktop VM (not the SIFT VM) back to the base snapshot and I installed the SIFT CLI on the fresh Ubuntu OS. Even in that case I still get the redis error when I try running L2T.

**Comment by joachimmetz** [comment]
This might be due to the Python version used by IPython, use `python3` or `ipython3` instead. Recent Plaso versions use your systems Python 3 version.

**Comment by honorthecode** [comment]
Thanks. I ran `ipython3` and the `redis.__file__` was `'/usr/local/lib/python3.6/dist-packages/redis/__init__.py'`. When I looked into that file or run `redis.__version__` I see that the redis version is 3.3.11
![image](https://user-images.githubusercontent.com/69635449/91663068-85525600-eab4-11ea-8bca-95fbb9b46aad.png)

I'm going to start researching how I can upgrade to a redis 3.4 or newer. I you know of a quick way to do so I would very much appreciate being pointed in the right direction.

Either way, thanks so much for the help so far!

**Comment by honorthecode** [comment]
As a side note I ran and `pip list` and `pip list | grep redis` and did not see redis in either.

**Comment by honorthecode** [comment]
Hmmm, I'm not seeing a version newer that 3.3.11 here- [https://tracker.debian.org/pkg/python-redis](https://launchpad.net/ubuntu/+source/python-redis), but I do see 3.5.3 here- [https://tracker.debian.org/pkg/python-redis](https://tracker.debian.org/pkg/python-redis).

I will give this a shot and see if I can figure out how to install it.

**Comment by honorthecode** [comment]
Progress! I updated the python3-redis and then was able to run log2timeline and then psort on a evtx file. I psortedthe output to json_line format and I can see that the for a 4624 event additional fields, such as the IP, SID, and username were parsed out. Te next step is to try psteal directly into an ELK stack.

![image](https://user-images.githubusercontent.com/69635449/91664454-67d5ba00-eabd-11ea-822b-2a5838517f41.png)

This is what I did:
1. Downloaded a newer version of redis from here: [https://tracker.debian.org/pkg/python-redis](https://tracker.debian.org/pkg/python-redis)
2. Scanned it with local AV and also uploaded to VirusTotal (no alerts on either)
3. Decompressed the new redis
4. cd into the redis package directory
5. `sudo python3 setup.py install`
6. Verify the new installed python3-redis version number using `ipython3` > `import redis` > `redis.__version__`
7. Test on a evtx file

**Comment by honorthecode** [comment]
Aaaaaand it broke again.... I ran `sudo apt-get install plaso-tools` to make sure plaso was as updated as possible and now I get the redis error event though ipython3 is showing me that the redis version is 3.5.3. Just to be safe I updated redis on python2 aswell and still get the redis version.


How can I update the version of redis that the GIFT PPA is installing and using?

**Comment by joachimmetz** [comment]
> As a side note I ran and pip list and pip list | grep redis and did not see redis in either.

again this is likely python2, try pip3

> I psortedthe output to json_line format and I can see that the for a 4624 event additional fields, such as the IP, SID, and username were parsed out. 

I don't think this is the latest version of Plaso

> How can I upgrade the version of redis that the GIFT PPA is installing and using?

Seeing that your local configuration is causing you a lot of issue, maybe use the Docker release?

**Comment by honorthecode** [comment]
Thank you for the pip3 suggestion. I did try out the Docker container for Plasoover the weekend and that did work just as expected. Since in the past I was able to get those other fields parsed out I was hoping I could do that again using the most recent version of Plaso. Then I would be able to do some nice analytics/charts using ELK780.

I was able to fix the host redis version using `pip3 install redis` and then was able to psteal data directly into ELK (and this is with the psteal version 20200717 installed on the host). Unfortunately  I do not have the additional fields parsed out but perhaps there is a way to parse out those fields using Kibana.

Thank you so much for all the responses @joachimmetz !
